File: 38-count-and-say/count-and-say.cpp
class Solution {
public:
    string countAndSay(int n) {
        vector<string> dp(n+1);

        dp[1] = "1";

        for(int i=2; i<=n; i++){
            string s = dp[i-1];
            string res;

            for(int j = 0; j<s.length(); j++){
                char c = s[j];
                int cnt = 0;

                while(j < s.length() && s[j] == c){
                    j++;
                    cnt++;
                }

                res += (to_string(cnt) + c);
                j--;
            }

            dp[i] = res;
        }

        return dp[n];
    }
};

File: 38-count-and-say/count-and-say.go
func helper(s string) string {
    var res string
    for i := 0; i < len(s); i++ {
        char := s[i]
        count := 0
        for i < len(s) && s[i] == char {
            count++
            i++
        }

        res += (toString(count) + string(char))
        i--
    }

    return res
}
func countAndSay(n int) string {
    dp := make([]string, n)
    
    dp[0] = "1"

    for i := 1; i < n; i++ {
        dp[i] = helper(dp[i-1])
    }
    return dp[n-1]
}
func toString(n int) string {
    return strconv.Itoa(n)
}